### PR TITLE
ENH: Add StringMethods.partition and rpartition

### PR DESCRIPTION
**Pull request by sinhrks**
Derived from #9111.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
What do you think of having `return_type` defaults to 'frame'? (I know in `str.split` the default is 'series', but that is for back compatibility)

**Comment by shoyer** [comment]
+1 for defaulting to return a frame.

**Comment by sinhrks** [comment]
OK, changed default to 'frame'.

**Comment by jreback** [comment]
lgtm, but not sure when would use ``partition`` over ``split``

**Comment by jorisvandenbossche** [comment]
`partition` will always give three items back (only split on first or last occurence of separator) I think?

@sinhrks maybe add some examples to the docstring, and put a 'see also' to split explaining the difference.

**Comment by sinhrks** [comment]
Added docstrings (examples and see also). `partition` is more specific than `split`, may useful when want to retain separators.

File: pandas/core/strings.py [hunk 1]
```diff
@@ -978,6 +982,64 @@ def split(self, pat=None, n=-1, return_type='series'):
...
+    Split the string at the %(side)s occurrence of sep, and return a 3-tuple containing the part
+    before the separator, the separator itself, and the part after the separator.
+    If the separator is not found, return %(return)s.
+
+    Parameters
+    ----------
+    pat : string, default whitespace
+        String to split on.
+    return_type : {'series', 'frame'}, default 'frame'
+        If frame, returns a DataFrame (elements are strings)
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: Small detail, but can you put a point `.` at the end of this sentence (as in html, the line breaks are not preserved, so the following `If series, ...` comes directly behind this, so without punctuation it looks strange

File: pandas/core/strings.py [hunk 2]
```diff
@@ -978,6 +982,64 @@ def split(self, pat=None, n=-1, return_type='series'):
...
+        If series, returns an Series (elements are lists of strings).
+
+    Returns
+    -------
+    split : array
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: arrays does not seems to be the right return type?

File: pandas/core/strings.py [hunk 3]
```diff
@@ -978,6 +982,64 @@ def split(self, pat=None, n=-1, return_type='series'):
...
+    0    A_B_C
+    1    D_E_F
+    2        X
+    dtype: object
+
+    >>> s.str.partition('_')
+       0  1    2
+    0  A  _  B_C
+    1  D  _  E_F
+    2  X
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: Do we want empty strings here, or NaNs? (which is more consistent with `split`)

Comment (jorisvandenbossche) [inline pandas/core/strings.py]: no matter, that is of course the behaviour of the standard library `partition`

Comment (sinhrks) [inline pandas/core/strings.py]: Yeah, it returns empty char based on standard ``partition``.

**Comment by sinhrks** [comment]
@jorisvandenbossche Thanks, fixed your points.

For returning type, I'll prepare separate PR for existing funcs which is also incorrect.

**Comment by sinhrks** [comment]
Hmm, ``StringMethods`` docstrings are mostly derived from corresponding module level functions, and there are 2 policies:

1. Describe the return type when called via ``StringMethods`` (describe what user will see)

  - https://github.com/pydata/pandas/blob/master/pandas/core/strings.py#L396

2. Describe the return type of the module level function (what the module level function returns)

  - https://github.com/pydata/pandas/blob/master/pandas/core/strings.py#L541

First policy looks preferable, so OK to change all the descriptions like first one? I understand that these module level functions are not public.

**Comment by jorisvandenbossche** [comment]
Yes, there is no real good solution to have it correct for both (unless starting with templating, but then the purpose of having a visual docstring you can correctly see in the source code for the module level functions is a bit lost). So I would say: most important that user sees the correct info, so I would indeed go with the first one.

**Comment by jorisvandenbossche** [comment]
@sinhrks or, maybe what we could also do is this: replace the array with Series, and leave the module level function docstrings as is. 
Eg for the example of `str_findall`: have a `.replace('matches : array', 'matches : Series')` on the docstring

BTW: it is not that there is a different policy in the docstrings apparently. Eg the `str_extract` you linked to *is* really returning already a Series/DataFrame. So it is just that the `str_` functions have different return types.

**Comment by sinhrks** [comment]
Separate docstring issue to #9843 as we have to fix other methods, and also consider #9667.

**Comment by jorisvandenbossche** [comment]
@sinhrks can you update this with `expand=True/false` instead of return_type? (see discussion in  #9870)

**Comment by sinhrks** [comment]
OK, fixed this. And posting the behavior of ``expand`` in #9870.

**Comment by jreback** [comment]
looks fine. needs a rebase.

File: pandas/core/strings.py [hunk 4]
```diff
@@ -1022,6 +1052,64 @@ def split(self, pat=None, n=-1, return_type='series'):
...
+    Split the string at the %(side)s occurrence of sep, and return a 3-tuple containing the part
+    before the separator, the separator itself, and the part after the separator.
+    If the separator is not found, return %(return)s.
+
+    Parameters
+    ----------
+    pat : string, default whitespace
+        String to split on.
+    expand : bool, default True
+        If True, return DataFrame/MultiIndex expanding dimensionality
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: Can you add a `.` at the end of the sentence.

Or make it a list:

```
* If true ..
* If false ..
```

File: pandas/core/strings.py [hunk 5]
```diff
@@ -1022,6 +1052,64 @@ def split(self, pat=None, n=-1, return_type='series'):
         result = str_split(self.series, pat, n=n, return_type=return_type)
         return self._wrap_result(result)
 
+    _shared_docs['str_partition'] = ("""
+    Split the string at the %(side)s occurrence of sep, and return a 3-tuple containing the part
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: `sep` -> `` `sep` ``

Comment (jorisvandenbossche) [inline pandas/core/strings.py]: Other thing: here you explain the '3-tuple' as output, but actually the three columns (expand=True) is the default

File: pandas/core/strings.py [hunk 6]
```diff
@@ -1012,6 +1014,34 @@ def _wrap_result(self, result):
...
+
+        if isinstance(self.series, Index):
+            name = getattr(result, 'name', None)
+            # if result is a boolean np.array, return the np.array
+            # instead of wrapping it into a boolean Index (GH 8875)
+            if hasattr(result, 'dtype') and is_bool_dtype(result):
+                return result
+
+            if expand:
+                return Index(list(result), name=name)
```
Comment (jreback) [inline pandas/core/strings.py]:
```
if expand:
    result = list(result)
return Index(result, name=name)
```

**Comment by sinhrks** [comment]
@jreback @jorisvandenbossche Thanks, fixed your points. Pls check again.

**Comment by jorisvandenbossche** [comment]
Looks good!

**Pull request merged**